### PR TITLE
Support SoftAffinity Scheduler to FileScanRDD

### PR DESCRIPTION
**Pull request by zzcclp**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/softaffinity/SoftAffinityManager.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  def usingSoftAffinity(): Boolean = _usingSoftAffinity
+
+  def handleExecutorAdded(execHostId: (String, String)): Unit = {
+    resourceRWLock.writeLock().lock()
+    try {
+      val executorsSet = nodesExecutorsMap.getOrElseUpdate(execHostId._2,
+        new mutable.HashSet[String]())
+      executorsSet.add(execHostId._1)
+      if (fixedIdForExecutors.exists(_.isEmpty)) {
```
Comment (hit-lacus) [inline core/src/main/scala/org/apache/spark/softaffinity/SoftAffinityManager.scala:64]: `fixedIdForExecutors.exists(_.isEmpty)` 表达的含义是什么？

Comment (zzcclp) [reply core/src/main/scala/org/apache/spark/softaffinity/SoftAffinityManager.scala:64]: 判断fixedIdForExecutors中是否有remove掉的executors，如果有，就是用新增的executor设置到当前位置，避免频繁变化fixedIdForExecutors的size

**Comment by zzcclp** [comment]
retest this please

**Comment by zzcclp** [comment]
retest this please

**Comment by zzcclp** [comment]
retest this please

**Comment by zzcclp** [comment]
retest this please

**Pull request closed without merging**